### PR TITLE
Fixes #2219 - adding a subcommand to enable deleting parts of a repo in the local cache

### LINKED ISSUE #2219: Select which files to delete in `huggingface-cli delete-cache`
**Is your feature request related to a problem? Please describe.**
The download tools all either dynamically or configurably download parts of a repo, since there are many times when the repo contains more files than needed (e.g. .pt, .safetensors). Once they're downloaded, you can't delete only part of the repo from your cache, and if you accidentally (or because you didn't know what you were doing) downloaded huge volumes of files, you can't delete just a part of it. The only option right now is to delete the whole repo and then re-download only what you want in your cache. 

**Describe the solution you'd like**
Option to select files within a given cached repo, using the existing tools, that are to be deleted. 

**Describe alternatives you've considered**
Manually manipulating files in the cache has gone..let's say "poorly" in the past. I've had to scrap my entire cache more than once doing that. 

**Additional context**
n/a

### PR DESCRIPTION
**Pull request by sealad886**
Initial thoughts: 

Going with `repo-cache`, use case would be something like:
```bash
> huggingface-cli repo-cache delete --all repo_id     # same functionality as delete-cache but quick to delete just the one repo
> huggingface-cli repo-cache delete --include (glob) repo_id.  # use same glob format as download --include flag; mark files to __include__ for delete action
> huggingface-cli repo-cache delete --exclude [glob] repo_id. # same as --include, but opposite
> huggingface-cli repo-cache delete --file-list (list of files to remove) repo_id
```

Also, a starting point is that the code references a frozen class `DeletesCacheStrategy`. I will do more digging, but there may be a framework that needs a little polish to make this very straightforward. Dangerous words in this business, I tell ya.

### CODE REVIEW

**Comment by Wauplin** [comment]
Hi @sealad886, thanks for the proposal and sorry for the long delay! So I've looked back to your snippet (and suggestions in https://github.com/huggingface/huggingface_hub/issues/2219#issuecomment-2050055659) and I think we could modify it a bit. What do you it of:

```
> huggingface-cli cache delete-repo repo_id
> huggingface-cli cache delete-repo repo_id --repo-type=dataset
> huggingface-cli cache delete-repo repo_id --include (glob)
> huggingface-cli cache delete-repo repo_id --exclude (glob)
> huggingface-cli cache delete-repo repo_id --files=file1,file2,file3,file4
```
?

Let's not focus too much on what's existing in `huggingface-cli delete-cache` BTW. I would prefer to have a unified experience to delete things from the cache but that's not really possible while maintaining an intuitive API. At least we can start by having a subcommands in a generic `huggingface-cli cache` part.

**Comment by sealad886** [comment]
@Wauplin those look like excellent starts. I'll probably also includes a `--dirs` (or if there's another naming convention used elsewhere in the repo, I'll rename it) to be explicit about deleting whole directories. 

I had been looking at some of the `git -i` stuff only recently, and it looks a lot like a utility that I've been developing for Ollama CLI (also centering on cache management). (Ollamaultil)[https://github.com/sealad886/ollama_util] is my take on a interface-based cache management utility (different focus: I use it to move cache files between on-disk and external storage--and back--since I'm on a laptop); it's not really adapted for components to be called at the command line, but it could be. Just wanted to throw that out there.

In terms of design, I think I'm going to start by actually enhancing the DeleteCacheStrategy class as the helper tool it is, and then working up from there. I've also noticed that the directories can sometimes fail to be deleted if (laughs with this one) '.DS_Store' files exist in the directory that's being deleted, and the delete process overall fails mid-way through. I'll fix this too. 

Also...verbosity.

So, to summarize my clearly scatterbrained thoughts:
1. CLI command `delete-repo` and associated options as noted above. 
2. Fix .DS_Store or other hidden files preventing the final deletion of a directory when the whole repo is being deleted.
3. Add verbosity, for now only to give some concept of progress when deleting large amounts of cached data.
4. Possibility of some kind of utility in the future?

I have time this week to actually dig into this more, so I'll get crackin'.

**Comment by Wauplin** [comment]
Thanks for writing up your thoughts @sealad886! This looks like a good plan to me. Please let me know if yo u need any assistance :)
